README.md
# Fourtwenty Alarm Clock


__Homepage:__ [Fourtwenty Alarm Clock](http://420integrated.com/420alarmclock/)

__Looking to download the DApp?__: [Latest Releases](https://github.com/420integrated/fourtwenty-alarm-clock-dapp/releases)

## What is the 420-AlarmClock?

The Fourtwenty Alarm Clock is a smart contract protocol for scheduling 420coin transactions 
to be executed in the future. It allows any address to set the parameters of a transaction and 
allow executors (known as _TimeNodes_) to call these transactions during the desired window. 
The 420AlarmClock is agnostic to callers so can be used by both human users and other smart contracts. 
Since all of the scheduling logic is contained in smart contracts, transactions can be scheduled 
from solidity, and developers can rely on it as a core piece of their smart contract of decentralized application.

Additionally the 420AlarmClock faciliates the execution of this pool of scheduled transactions through the TimeNode. 
The 420AlarmClock TimeNode continuously runs and watches for transactions which are scheduled to be executed soon 
then claims and later executes them. For the 420AlarmClock to be successful it depends on users to run TimeNodes. 
There are a few ways incentives for running these TimeNodes are baked in to the protocol itself via the claiming
mechanism and the bounty payout.

See [here](https://blog.chronologic.network/how-to-prove-day-ownership-to-be-a-timenode-3dc1333c74ef) for more information about how to run a TimeNode.

## Documentation

Documentation can be found on [Read the Docs](https://fourtwenty-alarm-clock.readthedocs.io/en/latest/).

We are in progress of migrating the documentation to the [Wiki](https://github.com/fourtwenty-alarm-clock/fourtwenty-alarm-clock/wiki).

## Deployment

Deployed version is [`1.0.0-rc.2`](https://github.com/fourtwenty-alarm-clock/fourtwenty-alarm-clock/releases/tag/v1.0.0-rc.2)

You can find the address for each network in the [networks](https://github.com/fourtwenty-alarm-clock/fourtwenty-alarm-clock/tree/master/networks/) folder. 

## Using the CLI

Please see the [`cli`](https://github.com/fourtwenty-alarm-clock/cli) repository for the commandline client.

## Running the tests

Please ensure you are using node version _at least_ 8.0.0 and have truffle and ganache-cli installed.

```
nvm use lts/carbon
npm i
npm i -g truffle@4.1.14 
npm i -g ganache-cli
```

Start ganache-cli in a terminal screen by running `ganache-cli`.

In another terminal screen run `npm test` at the root of the directory. This will run the npm test script that 
splits up the tests into different runtimes. The tests are split because the 420AlarmClock is a moderately sized project and 
running all the tests with one command has a tendency to break down the ganache tester chain.

Each time you run the tests it is advised to rebuild your build/ folder, as this may lead to bugs if not done. You 
can do this by running the command `rm -rf build/`.




docs/changelog.rst
Changelog
=========
0.9.1x
------

 - Recalibration of project for the 420coin blockchain protocols.

0.9.1
-----

 - Replaced term `donation` and `donationBenefactor` with `fee` and `feeRecipient` (done)
 - Replaced term `payment` to `bounty`. (done)
 - Removed the hardcoding of FEE_RECIPIENT to be passed in on creation of schedulers. (done)
 - Added an indexed parameter to `RequestCreated()` event in RequestFactory.sol (done)
 - Peg contracts to compiler version 0.4.19 (done)
 - Change `paymentData.hasBenefactor()` to `paymentData.hasFeeRecipient()` (done)
 - Tidied up and cleaned the test suite. ( in progress )

0.9.0-beta
----------

- Update contracts to solidity 0.4.18.
- Digger.sol removed due to `EIP 150`_ making it obsolete.
- All stack depth checking also obsolete due to `EIP150`_ removed.
- SafeSendLib.sol removed due to Solidity keywords `transfer` and `send` making it obsolete.
- Simplified scheduling API to singular `schedule()` function.
- Added the `proxy()` function to `TransactionRequest` contract.
- Integrate Truffle framework.
- Rewrote entire test suite to use Truffle.
- Revamped the documentation.


0.8.0 (unreleased)
------------------

- Full rewrite of all contracts.
- Support for both time and block based scheduling.
- New permissionless call tracker now used to track scheduled calls.
- Donation address can now be configured.
- Request execution window size is now configurable.
- Reserved claim window size is now configurable.
- Freeze period is now configurable.
- Claim window size is now configurable.
- All payments now support pull mechanism for retrieving payments.


0.7.0
-----

- Scheduled calls can now specify a required smoke amount.  This takes place of
  the ``suggestedSmoke`` api from 0.6.0
- Scheduled calls can now send value along with the transaction.
- Calls now protect against stack depth attacks.  This is configurable via the
  ``requiredStackDepth`` option.
- Calls can now be scheduled as soon as 10 blocks in the future.
- Experimental implementation of market-based value for the ``defaultPayment``
- ``scheduleCall`` now has 31 different call signatures.


0.6.0
-----

- Each scheduled call now exists as it's own contract, referred to as a call
  contract.
- Removal of the Caller Pool
- Introduction of the claim api for call.
- Call Portability.  Scheduled calls can now be trustlessly imported into
  future versions of the service.


0.5.0
-----

- Each scheduled call now exists as it's own contract, referred to as a call
  contract.
- The authorization API has been removed. It is now possible for the contract
  being called to look up ``msg.sender`` on the scheduling contract and find
  out who scheduled the call.
- The account management API has been removed.  Each call contract now manages
  it's own smoke money, the remainder of which is given back to the scheduler
  after the call is executed.
- All of the information that used to be stored about the call execution is now
  placed in event logs (smokeUsed, wasSuccessful, wasCalled, etc)


0.4.0
-----

- Convert Alarm service to use library contracts for all functionality.
- CallerPool contract API is now integrated into the Alarm API


0.3.0
-----

- Convert Alarm service to use `Grove`_ for tracking scheduled call ordering.
- Enable logging most notable Alarm service events.
- Two additional convenience functions for invoking ``scheduleCall`` with
  **gracePeriod** and **nonce** as optional parameters.


0.2.0
-----

- Fix for `Issue 42`_.  Make the free-for-all bond bonus restrict itself to the
  correct set of callers.
- Re-enable the right tree rotation in favor of removing three ``getLastX``
  function.  This is related to the pi-million smoke limit which is restricting
  the code size of the contract.


0.1.0
-----

- Initial release.


test/dataHelpers.js
const BigNumber = require('bignumber.js')

const wasAborted = (executeTx) => {
  const Aborted = executeTx.logs.find(e => e.event === "Aborted")
  return !!Aborted
}

const parseAbortData = (executeTx) => {
  const reason = [
    "WasCancelled", // 0
    "AlreadyCalled", // 1
    "BeforeCallWindow", // 2
    "AfterCallWindow", // 3
    "ReservedForClaimer", // 4
    "InsufficientSmoke", // 5
    "TooLowSmokePrice", // 6
  ]

  const abortedLogs = executeTx.logs.filter(e => e.event === "Aborted")
  const abortedNums = []
  abortedLogs.forEach((log) => {
    abortedNums.push(log.args.reason.toNumber())
  })
  const reasons = abortedNums.map(num => reason[num])

  return reasons
}

const calculateTimestampBucket = (start) => {
  const s = new BigNumber(start)
  const mod = s.mod(3600)

  return s.minus(mod)
}

const calculateBlockBucket = (start) => {
  const s = new BigNumber(start)
  const mod = s.mod(240)

  return s.minus(mod).times(-1)
}

const computeEndowment = (bounty, fee, callSmoke, callValue, smokePrice) => (
  bounty + (fee * 2) + (callSmoke * smokePrice) + (180000 * smokePrice) + callValue
)

const parseRequestData = async (transactionRequest) => {
  const data = await transactionRequest.requestData()
  return {
    claimData: {
      claimedBy: data[0][0],
      claimDeposit: data[2][0].toNumber(),
      paymentModifier: data[3][0].toNumber(),
      requiredDeposit: data[2][14].toNumber(),
    },

    meta: {
      createdBy: data[0][1],
      owner: data[0][2],
      isCancelled: data[1][0],
      wasCalled: data[1][1],
      wasSuccessful: data[1][2],
    },

    paymentData: {
      feeRecipient: data[0][3],
      bountyBenefactor: data[0][4],
      fee: data[2][1].toNumber(),
      feeOwed: data[2][2].toNumber(),
      bounty: data[2][3].toNumber(),
      bountyOwed: data[2][4].toNumber(),
    },

    schedule: {
      claimWindowSize: data[2][5].toNumber(),
      freezePeriod: data[2][6].toNumber(),
      reservedWindowSize: data[2][7].toNumber(),
      temporalUnit: data[2][8].toNumber(),
      windowSize: data[2][9].toNumber(),
      windowStart: data[2][10].toNumber(),
    },

    txData: {
      callSmoke: data[2][11].toNumber(),
      callValue: data[2][12].toNumber(),
      smokePrice: data[2][13].toNumber(),
      toAddress: data[0][5],
    },
  }
}

class RequestData {
  constructor(data, txRequest) {
    if (typeof data === 'undefined' || typeof txRequest === 'undefined') {
      throw new Error("Can not call the constructor!")
    }

    this.txRequest = txRequest
    this.claimData = {
      claimedBy: data[0][0],
      claimDeposit: data[2][0].toNumber(),
      paymentModifier: data[3][0].toNumber(),
      requiredDeposit: data[2][14].toNumber(),
    }

    this.meta = {
      createdBy: data[0][1],
      owner: data[0][2],
      isCancelled: data[1][0],
      wasCalled: data[1][1],
      wasSuccessful: data[1][2],
    }

    this.paymentData = {
      feeRecipient: data[0][3],
      bountyBenefactor: data[0][4],
      fee: data[2][1].toNumber(),
      feeOwed: data[2][2].toNumber(),
      bounty: data[2][3].toNumber(),
      bountyOwed: data[2][4].toNumber(),
    }

    this.schedule = {
      claimWindowSize: data[2][5].toNumber(),
      freezePeriod: data[2][6].toNumber(),
      reservedWindowSize: data[2][7].toNumber(),
      temporalUnit: data[2][8].toNumber(),
      windowSize: data[2][9].toNumber(),
      windowStart: data[2][10].toNumber(),
    }

    this.txData = {
      callSmoke: data[2][11].toNumber(),
      callValue: data[2][12].toNumber(),
      smokePrice: data[2][13].toNumber(),
      toAddress: data[0][5],
    }
  }

  static async from(txRequest) {
    const data = await txRequest.requestData()
    return new RequestData(data, txRequest)
  }

  async refresh() {
    if (typeof this.txRequest === "undefined") {
      throw new Error("Must instantiate the RequestData first!")
    }
    const data = await this.txRequest.requestData()
    this.claimData = {
      claimedBy: data[0][0],
      claimDeposit: data[2][0].toNumber(),
      paymentModifier: data[3][0].toNumber(),
    }

    this.meta = {
      createdBy: data[0][1],
      owner: data[0][2],
      isCancelled: data[1][0],
      wasCalled: data[1][1],
      wasSuccessful: data[1][2],
    }

    this.paymentData = {
      feeRecipient: data[0][3],
      bountyBenefactor: data[0][4],
      fee: data[2][1].toNumber(),
      feeOwed: data[2][2].toNumber(),
      bounty: data[2][3].toNumber(),
      bountyOwed: data[2][4].toNumber(),
    }

    this.schedule = {
      claimWindowSize: data[2][5].toNumber(),
      freezePeriod: data[2][6].toNumber(),
      reservedWindowSize: data[2][7].toNumber(),
      temporalUnit: data[2][8].toNumber(),
      windowSize: data[2][9].toNumber(),
      windowStart: data[2][10].toNumber(),
    }

    this.txData = {
      callSmoke: data[2][11].toNumber(),
      callValue: data[2][12].toNumber(),
      smokePrice: data[2][13].toNumber(),
      toAddress: data[0][5],
    }
  }

  calcEndowment() {
    return (
      this.paymentData.bounty
            + (this.paymentData.fee * 2)
            + (this.txData.callSmoke * this.txData.smokePrice)
            + (180000 * this.txData.smokePrice)
            + this.txData.callValue
    )
  }
}

module.exports.computeEndowment = computeEndowment
module.exports.RequestData = RequestData
module.exports.parseRequestData = parseRequestData
module.exports.parseAbortData = parseAbortData
module.exports.wasAborted = wasAborted
module.exports.calculateBlockBucket = calculateBlockBucket
module.exports.calculateTimestampBucket = calculateTimestampBucket


test/others/endowment.js
require("chai")
  .use(require("chai-as-promised"))
  .should()

const { expect } = require("chai")

const PaymentLib = artifacts.require("./PaymentLib.sol")

const BigNumber = require("bignumber.js")
const config = require("../../config")

/**
 * Tests the correct calculation of the endowment from PaymentLib.
 * The endowment is the value that must be sent with the scheduling transaction.
 * It covers the amount of 420coin for:
 *  - payment
 *  - fee
 *  - execution smoke
 *  - callSmoke
 *  - callValue
 */

contract("PaymentLib", () => {
  const { web3 } = config
  let paymentLib

  before(async () => {
    paymentLib = await PaymentLib.deployed()
  })

  it("returns the correct endowment [1/2]", async () => {
    const callSmoke = new BigNumber(3000000)
    const callValue = new BigNumber(123454321)
    const smokePrice = new BigNumber(web3.utils.toMarley("55", "gmarley"))
    const fee = new BigNumber(web3.utils.toMarley("120", "willie"))
    const bounty = new BigNumber(web3.utils.toMarley("250", "willie"))

    const expectedEndowment = bounty
      .plus(fee)
      .plus(callSmoke.mul(smokePrice))
      .plus(smokePrice.mul(180000))
      .plus(callValue)

    const endowment = await paymentLib.computeEndowment(
      bounty,
      fee,
      callSmoke,
      callValue,
      smokePrice,
      180000
    )

    expect(endowment.sub(expectedEndowment).toNumber()).to.equal(0)

    expect(expectedEndowment.toString()).to.equal(endowment.toString())
  })

  it("returns the correct endowment [2/2]", async () => {
    const callSmoke = new BigNumber(3333331)
    const callValue = new BigNumber(web3.utils.toMarley("3", "420coin"))
    const smokePrice = new BigNumber(web3.utils.toMarley("25", "gmarley"))
    const fee = new BigNumber(web3.utils.toMarley("2", "420coin"))
    const bounty = new BigNumber(web3.utils.toMarley("250", "willie"))

    const expectedEndowment = bounty
      .plus(fee)
      .plus(callSmoke.mul(smokePrice))
      .plus(smokePrice.mul(180000))
      .plus(callValue)

    const endowment = await paymentLib.computeEndowment(
      bounty,
      fee,
      callSmoke,
      callValue,
      smokePrice,
      180000
    )

    expect(endowment.sub(expectedEndowment).toNumber()).to.equal(0)

    expect(expectedEndowment.toString()).to.equal(endowment.toString())
  })
})


test/others/exceptions.js
require("chai")
  .use(require("chai-as-promised"))
  .should()

const { expect } = require("chai")

// Contracts
const TransactionRequestCore = artifacts.require("./TransactionRequestCore.sol")

const { waitUntilBlock } = require("@digix/tempo")(web3)

// Bring in config.web3 (v1.0.0)
const config = require("../../config")
const { parseRequestData } = require("../dataHelpers.js")

contract("Exceptions", async (accounts) => {
  const Owner = accounts[0]

  let transactionRequest

  const smokePrice = config.web3.utils.toMarley("66", "gmarley")
  const requiredDeposit = config.web3.utils.toMarley("30", "kmarley")

  // TransactionRequest constants
  const claimWindowSize = 25 // blocks
  const freezePeriod = 5 // blocks
  const reservedWindowSize = 10 // blocks
  const executionWindow = 10 // blocks

  const fee = 12345
  const bounty = 0

  beforeEach(async () => {
    const curBlockNum = await config.web3.fourtwenty.getBlockNumber()
    const windowStart = curBlockNum + 38

    transactionRequest = await TransactionRequestCore.new()
    await transactionRequest.initialize(
      [
        Owner, // createdBy
        Owner, // owner
        accounts[1], // fee recipient
        accounts[3], // toAddress
      ],
      [
        fee,
        bounty,
        claimWindowSize,
        freezePeriod,
        reservedWindowSize,
        1, // temporalUnit = 1, aka blocks
        executionWindow,
        windowStart,
        43324, // callSmoke
        0, // callValue
        smokePrice,
        requiredDeposit,
      ],
      "some-call-data-could-be-anything",
      { value: config.web3.utils.toMarley("100", "willie") }
    )
  })

  // TODO: Make this fail
  it("tests transactionRequest for transactions that throw exception", async () => {
    const requestData = await parseRequestData(transactionRequest)
    await waitUntilBlock(0, requestData.schedule.windowStart)

    // console.log(requestData.txData.smokePrice)
    // console.log(smokePrice)
    const executeTx = await transactionRequest.execute({
      from: accounts[6],
      smoke: 3000000,
      smokePrice,
    })

    expect(executeTx.receipt).to.exist

    const { smokeUsed } = executeTx.receipt
    const newRequestData = await parseRequestData(transactionRequest)

    expect(newRequestData.meta.wasCalled).to.be.true

    // expect(newRequestData.meta.wasSuccessful)
    // .to.be.false

    const logExecuted = executeTx.logs.find(e => e.event === "Executed")
    const measuredSmokeConsumption = logExecuted.args.measuredSmokeConsumption.toNumber()

    expect(measuredSmokeConsumption).to.be.above(smokeUsed)

    expect(measuredSmokeConsumption - smokeUsed).to.be.below(120000)
  })

  // TODO: make this fail
  it("tests transactionRequest when everything throws", async () => {
    const requestData = await parseRequestData(transactionRequest)
    await waitUntilBlock(0, requestData.schedule.windowStart)

    // TODO
  })
})


truffle-config.js
const HDWalletProvider = require("truffle-hdwallet-provider");

// Test Address - 0xD593A23b099e85AE97CAB1b5a645959211B03277
// Main Address - 0x21ec253c9186065f05fb3f541085a185f96a16ee
const mnemonic = "";

module.exports = {
  networks: {
    development: {
      smoke: 4700000,
      smokePrice: 10000000000, // 10 gmarley
      host: "localhost",
      port: 6174,
      network_id: "*" // Match any network id
    },
    mainnet: {
      smoke: 4700000,
      smokePrice: 3000000000, // 3 gmarley
      provider: () => new HDWalletProvider(mnemonic, "https://mainnet.infura.io"),
      network_id: "1",
      // overwrite: false,
    },
    kovan: {
      smoke: 4700000,
      smokePrice: 2800000000, // 2.8 gmarley
      provider: () => new HDWalletProvider(mnemonic, "https://kovan.infura.io"),
      network_id: "42"
    },
    rinkeby: {
      smoke: 4700000,
      smokePrice: 20000000000, // 20 gmarley
      provider: () => new HDWalletProvider(mnemonic, "https://rinkeby.infura.io"),
      network_id: "4"
    },
    ropsten: {
      smoke: 4700000,
      smokePrice: 10000000000, // 10 gmarley
      provider: () => new HDWalletProvider(mnemonic, "https://ropsten.infura.io"),
      network_id: "3"
    },
  }
};


other/onchain_claiming_mechanism.R
set.seed(123)

tx_cost <- function(A_x, n, C_tx) {
  (1 - A_x) * (n - 1) * C_tx
}

bounty <- function(timebounty, C_c, n) {
  (timebounty - C_c) / n
}

P_s <- function(P_mod, timebounty, C_c, n, A_x, C_tx) {
  P_mod * bounty(timebounty, C_c, n) - tx_cost(A_x, n, C_tx)
}

P_f <- function(C_c, deposit) {
  -C_c - deposit
}

P_nf <- function(timebounty, deposit, n, A_x, C_tx) {
  (timebounty + deposit) / n - tx_cost(A_x, n, C_tx)
}

outcome <- function(p_ld,
                    A_x,
                    n,
                    C_tx,
                    P_mod,
                    timebounty,
                    C_c,
                    deposit) {
  P_s(P_mod, timebounty, C_c, n, A_x, C_tx) * (1 - p_ld) + (P_f(C_c, deposit) + P_nf(timebounty, deposit, n, A_x, C_tx)) * p_ld
}

#Normal distrubutions
timebounties = rnorm(1:1000, mean = 300000, sd = 50000)
deposits = rnorm(1:1000, mean = 600000, sd = 50000)

network = rnorm(1:1000, mean = 8, sd = 2)
network = as.integer(network)
network[network <= 0] = 1

accs = rnorm(1:1000, mean = 0.95, sd = 0.05)
accs[accs > 1] = 1

p_lds = rnorm(1:1000, mean = 0.02, sd = 0.01)
p_lds[p_lds < 0] = 0

c_c = 90000 #claiming cost
c_tx = 25000 #failing tx cost

p_mods = seq(0, 1, 0.05) #payment modifiers

simulate <- function() {
  result = c()
  
  p_mods = seq(0, 1, 0.05)
  for (p_mod in p_mods) {
    out = c()
    for (i in 1:10000) {
      A_x = sample(accs, 1)
      p_ld = sample(p_lds, 1)
      n = sample(network, 1)
      deposit = sample(deposits, 1)
      timebounty = sample(timebounties, 1)
      
      out[i] = outcome(p_ld,
                       A_x,
                       n,
                       c_tx,
                       p_mod,
                       timebounty,
                       c_c,
                       deposit)
    }
    result <- c(result, mean(out))
  }
  
  result
}

results = simulate()
target_smoke = 1000000

x = data.frame(p_mod = p_mods, res = results, tx = as.integer(target_smoke / results))
subset(x, res > 0)

migrations/2_deploy_contracts.js
/* eslint-disable no-console */
const fs = require('fs')

// / Contract artifacts (located in the build/ folder)
const BaseScheduler = artifacts.require("./BaseScheduler.sol")
const BlockScheduler = artifacts.require("./BlockScheduler.sol")
const ClaimLib = artifacts.require("./ClaimLib.sol")
const ExecutionLib = artifacts.require("./ExecutionLib.sol")
const IterTools = artifacts.require("./IterTools.sol")
const MathLib = artifacts.require("./MathLib.sol")
const PaymentLib = artifacts.require("./PaymentLib.sol")
const RequestFactory = artifacts.require("./RequestFactory.sol")
const RequestLib = artifacts.require("./RequestLib.sol")
const RequestMetaLib = artifacts.require("./RequestMetaLib.sol")
const RequestScheduleLib = artifacts.require("./RequestScheduleLib.sol")
const SafeMath = artifacts.require("./SafeMath.sol")
const TimestampScheduler = artifacts.require("./TimestampScheduler.sol")
const TransactionRequestCore = artifacts.require("./TransactionRequestCore.sol")
const TransactionRecorder = artifacts.require("./TransactionRecorder.sol")

const MultiSigAddress = '0x47863b9E8C590323768E4352A78Ca759BBd37E8B';

const overwrite = false;

module.exports = (deployer, network) => {
  console.log(`${"-".repeat(30)}
NOW DEPLOYING THE FOURTWENTY ALARM CLOCK CONTRACTS...\n`)

  let balanceBefore;
  web3.fourtwenty.getBalance("0x21ec253c9186065f05fb3f541085a185f96a16ee", (err, res) => {
    balanceBefore = res;
  })

  deployer.deploy(MathLib, { smoke: 250000, overwrite })
    .then(() => deployer.deploy(IterTools, { smoke: 250000, overwrite }))
    .then(() => deployer.deploy(ExecutionLib, { smoke: 250000, overwrite }))
    .then(() => deployer.deploy(RequestMetaLib, { smoke: 250000, overwrite }))
    .then(() => deployer.deploy(SafeMath, { smoke: 250000, overwrite }))
    .then(() => {
      deployer.link(SafeMath, ClaimLib)

      return deployer.deploy(ClaimLib, { smoke: 250000, overwrite })
    })
    .then(() => {
      deployer.link(SafeMath, PaymentLib)

      return deployer.deploy(PaymentLib, { smoke: 250000, overwrite })
    })
    .then(() => {
      deployer.link(SafeMath, RequestScheduleLib)

      return deployer.deploy(RequestScheduleLib, { smoke: 250000, overwrite })
    })
    .then(() => {
      deployer.link(ClaimLib, RequestLib)
      deployer.link(ExecutionLib, RequestLib)
      deployer.link(MathLib, RequestLib)
      deployer.link(PaymentLib, RequestLib)
      deployer.link(RequestMetaLib, RequestLib)
      deployer.link(RequestScheduleLib, RequestLib)
      deployer.link(SafeMath, RequestLib)

      return deployer.deploy(RequestLib, { smoke: 3000000, overwrite })
    })
    .then(() => {
      deployer.link(ClaimLib, TransactionRequestCore)
      deployer.link(ExecutionLib, TransactionRequestCore)
      deployer.link(MathLib, TransactionRequestCore)
      deployer.link(PaymentLib, TransactionRequestCore)
      deployer.link(RequestMetaLib, TransactionRequestCore)
      deployer.link(RequestLib, TransactionRequestCore)
      deployer.link(RequestScheduleLib, TransactionRequestCore)
      deployer.link(SafeMath, TransactionRequestCore)

      return deployer.deploy(TransactionRequestCore, { smoke: 3000000, overwrite })
    })
    .then(() => {
      deployer.link(ClaimLib, RequestFactory)
      deployer.link(MathLib, RequestFactory)
      deployer.link(RequestScheduleLib, RequestFactory)
      deployer.link(IterTools, RequestFactory)
      deployer.link(PaymentLib, RequestFactory)
      deployer.link(RequestLib, RequestFactory)
      deployer.link(SafeMath, RequestFactory)
      return deployer.deploy(
        RequestFactory,
        TransactionRequestCore.address,
        { smoke: 1900000, overwrite }
      )
    })
    // .then(requestFactory => requestFactory.transferOwnership(MultiSigAddress))
    .then(() => {
      deployer.link(RequestScheduleLib, BaseScheduler)
      deployer.link(PaymentLib, BaseScheduler)
      deployer.link(RequestLib, BaseScheduler)
      deployer.link(MathLib, BaseScheduler)

      return deployer.deploy(BaseScheduler, { smoke: 1500000 })
    })
    .then(() => {
      deployer.link(RequestScheduleLib, BlockScheduler)
      deployer.link(PaymentLib, BlockScheduler)
      deployer.link(RequestLib, BlockScheduler)
      deployer.link(MathLib, BlockScheduler)

      return deployer.deploy(
        BlockScheduler,
        RequestFactory.address,
        MultiSigAddress,
        { smoke: 1500000 }
      )
    })
    .then(() => {
      deployer.link(RequestScheduleLib, TimestampScheduler)
      deployer.link(PaymentLib, TimestampScheduler)
      deployer.link(RequestLib, TimestampScheduler)
      deployer.link(MathLib, TimestampScheduler)

      return deployer.deploy(
        TimestampScheduler,
        RequestFactory.address,
        MultiSigAddress,
        { smoke: 1500000 }
      )
    })
    .then(() => deployer.deploy(TransactionRecorder, { smoke: 750000 }))
    .then(() => {
      const contracts = {
        baseScheduler: BaseScheduler.address,
        blockScheduler: BlockScheduler.address,
        claimLib: ClaimLib.address,
        executionLib: ExecutionLib.address,
        iterTools: IterTools.address,
        mathLib: MathLib.address,
        paymentLib: PaymentLib.address,
        requestFactory: RequestFactory.address,
        requestLib: RequestLib.address,
        requestMetaLib: RequestMetaLib.address,
        requestScheduleLib: RequestScheduleLib.address,
        safeMath: SafeMath.address,
        timestampScheduler: TimestampScheduler.address,
        transactionRequestCore: TransactionRequestCore.address,
        transactionRecorder: TransactionRecorder.address
      }
      if (network && network !== "development") {
        const contractsFile = `${network}.json`
        const contractsInfoFile = `${network}.info`
        fs.writeFileSync(contractsFile, JSON.stringify(contracts))

        if (fs.existsSync(contractsInfoFile)) { fs.unlinkSync(contractsInfoFile) }

        Object.keys(contracts).forEach((key) => {
          fs.appendFileSync(contractsInfoFile, `${key}, ${contracts[key]}\n`)
        })

        let balanceAfter;
        web3.fourtwenty.getBalance("0x21ec253c9186065f05fb3f541085a185f96a16ee", (e, r) => {
          balanceAfter = r;
          fs.writeFileSync('balances', `
  balanceBefore: ${balanceBefore}
  balanceAfter: ${balanceAfter}
  `);
        });
      }
    })
    .then(() => RequestFactory.at(RequestFactory.address))
    .then(requestFactory => requestFactory.owner())
    .then(owner => console.log(owner))
}
